packages/dev/core/src/Meshes/Compression/dracoDecoder.ts

```diff
@@ -0,0 +1,282 @@
+import { type DecoderModule, DracoDecoderModule } from "draco3dgltf";
+import { DracoCodec, type IDracoCodecConfiguration } from "./dracoCodec";
+import { Tools } from "../../Misc/tools";
+import { Geometry } from "../geometry";
+import { VertexBuffer } from "../buffer";
+import { Logger } from "../../Misc/logger";
+import type { BoundingInfo } from "../../Culling/boundingInfo";
+import type { Scene } from "../../scene";
+import type { Nullable } from "../../types";
+import { decodeMesh, type AttributeData, type Message, workerFunction } from "./dracoCompressionWorker";
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+declare let DracoDecoderModule: DracoDecoderModule;
+
+interface MeshData {
+    indices?: Uint16Array | Uint32Array;
+    attributes: Array<AttributeData>;
+    totalVertices: number;
+}
+
+/**
+ * Draco compression (https://google.github.io/draco/)
+ *
+ * This class wraps the Draco decoder module.
+ *
+ * By default, the configuration points to a copy of the Draco decoder files for glTF from the Babylon.js preview cdn https://preview.babylonjs.com/draco_wasm_wrapper_gltf.js.
+ *
+ * To update the configuration, use the following code:
+ * ```javascript
+ *     DracoDecoder.Config = {
+ *          wasmUrl: "<url to the WebAssembly library>",
+ *          wasmBinaryUrl: "<url to the WebAssembly binary>",
+ *          fallbackUrl: "<url to the fallback JavaScript library>",
+ *     };
+ * ```
+ *
+ * Draco has two versions, one for WebAssembly and one for JavaScript. The decoder configuration can be set to only support WebAssembly or only support the JavaScript version.
+ * Decoding will automatically fallback to the JavaScript version if WebAssembly version is not configured or if WebAssembly is not supported by the browser.
+ * Use `DracoDecoder.Available` to determine if the decoder configuration is available for the current context.
+ *
+ * To decode Draco compressed data, get the default DracoDecoder object and call decodeMeshToGeometryAsync:
+ * ```javascript
+ *     var geometry = await DracoDecoder.Default.decodeMeshToGeometryAsync(data);
+ * ```
+ */
+export class DracoDecoder extends DracoCodec<DecoderModule> {
+    protected static override _Default: Nullable<DracoDecoder> = null;
+    /**
+     * Default instance for the DracoDecoder.
+     */
+    public static get Default(): DracoDecoder {

```

Comment:
I find it a bit confusing that there is both a static default instance, and also the class can be constructed with a custom config. It seems like we could shrink the API surface area and make it more clear by just having a default config and always requiring construction. If we go this route, I think we could simplify further by making the constructor private, introducing an async factory function, and removing `whenReadyAsync`. The usage patter would then by:
```ts
const defaultDecoder = await DracoDecoder.CreateAsync();
const customDecoder = await DracoDecoder.CreateAsync(myCustomConfig);
```

Reply:
@bghgary Do you have any opinion about making this change?

Reply:
We don't want the usage pattern to be create multiple instances. This class is heavy. It creates WebWorker pool, loads WebAssembly, etc. Typical usage should be there is only one config so that only one set of Draco assets are loaded.

I did it this way for the original DracoCompression class because the Draco extension for glTF loader needs to use Draco. We don't want to create multiple instances of the DracoCompression class. Maybe there is another way to solve it, but this is what I did originally.

Reply:
If typically there would just be one config, then maybe we can just cache and re-use either the DracoDecoder instance or the worker pool internally? Maybe with a WeakMap or a WeakRef so it can also be released when no longer being used (if it is heavy as you mention)? It seems like there should be a number of ways to make this complexity an implementation detail and keep the API simple (especially avoiding the `whenReadyAsync` so you can never have a partially initialized object in an invalid state).